core/src/test/scala/org/apache/spark/deploy/history/HistoryServerSuite.scala

```diff
@@ -183,19 +183,43 @@ class HistoryServerSuite extends SparkFunSuite with BeforeAndAfter with Matchers
   // in the test resource folder
   cases.foreach { case (name, path) =>
     test(name) {
-      val (code, jsonOpt, errOpt) = getContentAndCode(path)
-      code should be (HttpServletResponse.SC_OK)
-      jsonOpt should be ('defined)
-      errOpt should be (None)
-
-      val exp = IOUtils.toString(new FileInputStream(
-        new File(expRoot, HistoryServerSuite.sanitizePath(name) + "_expectation.json")))
-      // compare the ASTs so formatting differences don't cause failures
-      import org.json4s._
-      import org.json4s.jackson.JsonMethods._
-      val jsonAst = parse(clearLastUpdated(jsonOpt.get))
-      val expAst = parse(exp)
-      assertValidDataInJson(jsonAst, expAst)
+      val expectationFile = new File(expRoot, HistoryServerSuite.sanitizePath(name) +
+        "_expectation.json")
+      assertApiCallResponse(path, expectationFile)
+    }
+  }
+
+  val casesCustomLogUrl = Seq(
+    "completed app executor list json apply custom log urls" ->
+      "applications/application_1547723113049_0005/1/executors",
+    "incomplete app executor list json apply custom log urls" ->
+      "applications/application_1547723113049_0006/1/executors"
+  )
+
+  casesCustomLogUrl.foreach { case (name, path) =>
+    Seq(true, false).foreach { applyToIncompleteApp =>
+      val newName = name + s" apply incomplete app $applyToIncompleteApp"
+      test(newName) {
+        // This only verifies whether applying custom log URLs is in effect for SHS.
+        // Validation of "custom log URLs" functionality will be covered from different UTs.
+
+        // restart server
+        try {
+          stop()
+          init(
+            // compiler complains when using `->`, so made a tuple
+            (CUSTOM_EXECUTOR_LOG_URL.key, "http://newhost:9999/logs/clusters/" +
+              "{{CLUSTER_ID}}/users/{{USER}}/containers/{{CONTAINER_ID}}/{{FILE_NAME}}"),
```

Comment:
the compilation issue is probably that `->` has higher precedence, you can put parens around the string join:

```scala
scala> "foo" -> "a" + "b"
res0: String = (foo,a)b
scala> "foo" -> ("a" + "b")
res1: (String, String) = (foo,ab)
```

or change the second entry to also use parens -- I don't think it really matters, but nicer to have all entries here be consistent.